File: cpp/alerts/torrent_alert.cpp
#include "./torrent_alert.h"
#include "libtorrent/alert_types.hpp"
#include "libtorrent/fwd.hpp"
#include <memory>

namespace ltrs {
    std::unique_ptr<lt::torrent_handle> torrent_alert_handle(lt::torrent_alert *a) {
        return std::make_unique<lt::torrent_handle>(a->handle);
    }
    rust::String torrent_alert_message(const lt::torrent_alert *a) {
        return a->message();
    }
    rust::Str torrent_alert_torrent_name(const lt::torrent_alert *a) {
        return a->torrent_name();
    }
}

File: src/ffi/alerts/torrent_alert.rs
#[cxx::bridge(namespace = "ltrs")]
pub(crate) mod ffi {
    #[namespace = "libtorrent"]
    extern "C++" {
        type torrent_alert = crate::ffi::ffi::torrent_alert;
        type torrent_handle = crate::ffi::torrent_handle::ffi::torrent_handle;
    }
    unsafe extern "C++" {
        include!("cpp/alerts/torrent_alert.h");
        unsafe fn torrent_alert_handle(alert: *mut torrent_alert) -> UniquePtr<torrent_handle>;
        unsafe fn torrent_alert_message(alert: *const torrent_alert) -> String;
        unsafe fn torrent_alert_torrent_name<'a>(alert: *const torrent_alert) -> &'a str;
    }
}
